File: android2/app/src/main/java/com/xbot/xbot/base/BaseService.java
package com.xbot.xbot.base;

import android.annotation.SuppressLint;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.google.gson.Gson;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Classic Bluetooth SPP (RFCOMM) base controller service.
 */
public class BaseService {
    private static final String TAG = "BaseService";
    /** Standard SPP UUID (Serial Port Profile). */
    public static final UUID SPP_UUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB");
    private static final Gson GSON = new Gson();

    public enum ConnectionState {
        DISCONNECTED, CONNECTING, CONNECTED, ERROR
    }

    public static class BaseStatus {
        public int dof = 2;
        public String mode = "idle";
        public double yaw;
        public double pitch;
        public double roll;

        public BaseStatus copyWith(
                @Nullable Integer dof,
                @Nullable String mode,
                @Nullable Double yaw,
                @Nullable Double pitch,
                @Nullable Double roll) {
            BaseStatus s = new BaseStatus();
            s.dof = dof != null ? dof : this.dof;
            s.mode = mode != null ? mode : this.mode;
            s.yaw = yaw != null ? yaw : this.yaw;
            s.pitch = pitch != null ? pitch : this.pitch;
            s.roll = roll != null ? roll : this.roll;
            return s;
        }
    }

    private final BaseProtocol protocol = new BaseProtocol();
    private final ExecutorService io = Executors.newSingleThreadExecutor();
    private final MutableLiveData<ConnectionState> connectionState =
            new MutableLiveData<>(ConnectionState.DISCONNECTED);
    private final MutableLiveData<BaseStatus> status = new MutableLiveData<>(new BaseStatus());
    private final MutableLiveData<String> errorMessage = new MutableLiveData<>();
    private final AtomicBoolean readerRunning = new AtomicBoolean(false);

    @Nullable private BluetoothSocket socket;
    @Nullable private InputStream input;
    @Nullable private OutputStream output;
    @Nullable private BluetoothDevice device;
    @Nullable private FrameListener frameListener;

    public interface FrameListener {
        void onFrame(Map<String, Object> frame);
    }

    public LiveData<ConnectionState> getConnectionState() {
        return connectionState;
    }

    public LiveData<BaseStatus> getStatus() {
        return status;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    @Nullable
    public BluetoothDevice getDevice() {
        return device;
    }

    public boolean isConnected() {
        return connectionState.getValue() == ConnectionState.CONNECTED;
    }

    public void setFrameListener(@Nullable FrameListener listener) {
        this.frameListener = listener;
    }

    @SuppressLint("MissingPermission")
    public List<BluetoothDevice> getBondedDevices() {
        BluetoothAdapter adapter = BluetoothAdapter.getDefaultAdapter();
        if (adapter == null) {
            return new ArrayList<>();
        }
        Set<BluetoothDevice> bonded = adapter.getBondedDevices();
        return bonded != null ? new ArrayList<>(bonded) : new ArrayList<>();
    }

    @SuppressLint("MissingPermission")
    public void connect(BluetoothDevice target) {
        if (connectionState.getValue() == ConnectionState.CONNECTING) {
            return;
        }
        device = target;
        connectionState.postValue(ConnectionState.CONNECTING);
        io.execute(() -> {
            try {
                disconnectInternal("reconnect");
                BluetoothSocket s = target.createRfcommSocketToServiceRecord(SPP_UUID);
                s.connect();
                socket = s;
                input = s.getInputStream();
                output = s.getOutputStream();
                protocol.reset();
                connectionState.postValue(ConnectionState.CONNECTED);
                errorMessage.postValue(null);
                startReader();
                Log.i(TAG, "connected " + target.getAddress());
            } catch (IOException e) {
                Log.e(TAG, "connect failed", e);
                disconnectInternal("连接失败: " + e.getMessage());
            }
        });
    }

    public void disconnect() {
        io.execute(() -> disconnectInternal("主动断开"));
    }

    public boolean send(String cmd, Map<String, Object> params) {
        if (!isConnected() || output == null) {
            Log.w(TAG, "send while disconnected");
            return false;
        }
        byte[] bytes = protocol.encode(cmd, params);
        try {
            output.write(bytes);
            output.flush();
            return true;
        } catch (IOException e) {
            Log.e(TAG, "send failed", e);
            disconnectInternal("发送失败: " + e.getMessage());
            return false;
        }
    }

    private void startReader() {
        if (!readerRunning.compareAndSet(false, true)) {
            return;
        }
        io.execute(() -> {
            byte[] buf = new byte[4096];
            try {
                InputStream in = input;
                while (readerRunning.get() && in != null) {
                    int n = in.read(buf);
                    if (n < 0) {
                        break;
                    }
                    if (n == 0) {
                        continue;
                    }
                    List<Map<String, Object>> frames = protocol.feed(buf, 0, n);
                    for (Map<String, Object> frame : frames) {
                        handleFrame(frame);
                        FrameListener listener = frameListener;
                        if (listener != null) {
                            listener.onFrame(frame);
                        }
                        Log.d(TAG, "RX " + GSON.toJson(frame));
                    }
                }
            } catch (IOException e) {
                if (readerRunning.get()) {
                    Log.e(TAG, "read error", e);
                }
            } finally {
                readerRunning.set(false);
                disconnectInternal("设备断开");
            }
        });
    }

    private void handleFrame(Map<String, Object> frame) {
        Object cmdObj = frame.get("cmd");
        if (!(cmdObj instanceof String)) {
            return;
        }
        String cmd = (String) cmdObj;
        BaseStatus current = status.getValue();
        if (current == null) {
            current = new BaseStatus();
        }
        switch (cmd) {
            case "status":
                BaseStatus updated = new BaseStatus();
                updated.dof = intVal(frame.get("dof"), current.dof);
                updated.mode = stringVal(frame.get("mode"), current.mode);
                updated.yaw = numVal(frame.get("yaw"), current.yaw);
                updated.pitch = numVal(frame.get("pitch"), current.pitch);
                updated.roll = numVal(frame.get("roll"), current.roll);
                status.postValue(updated);
                break;
            case "version":
                Object dof = frame.get("dof");
                if (dof instanceof Number) {
                    status.postValue(current.copyWith(((Number) dof).intValue(), null, null, null, null));
                }
                break;
            default:
                break;
        }
    }

    private void disconnectInternal(String reason) {
        readerRunning.set(false);
        try {
            if (input != null) {
                input.close();
            }
        } catch (IOException ignored) {
        }
        try {
            if (output != null) {
                output.close();
            }
        } catch (IOException ignored) {
        }
        try {
            if (socket != null) {
                socket.close();
            }
        } catch (IOException ignored) {
        }
        input = null;
        output = null;
        socket = null;
        protocol.reset();
        connectionState.postValue(ConnectionState.DISCONNECTED);
        errorMessage.postValue(reason);
        Log.i(TAG, reason);
    }

    public void release() {
        disconnect();
        io.shutdownNow();
    }

    private static int intVal(@Nullable Object v, int fallback) {
        return v instanceof Number ? ((Number) v).intValue() : fallback;
    }

    private static double numVal(@Nullable Object v, double fallback) {
        return v instanceof Number ? ((Number) v).doubleValue() : fallback;
    }

    private static String stringVal(@Nullable Object v, String fallback) {
        return v instanceof String ? (String) v : fallback;
    }
}

File: xbot-native/app/src/main/java/com/xbot/android/base/BaseService.kt
package com.xbot.android.base

import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothSocket
import android.content.Context
import android.util.Log
import org.json.JSONObject
import java.io.IOException
import java.util.UUID

/**
 * 底座控制服务：经典蓝牙 SPP(RFCOMM) 连接管理 + JSON 指令收发。
 * 对应 Flutter base_service.dart。
 *
 * 标准 SPP UUID 00001101-0000-1000-8000-00805F9B34AF。
 *
 * 对外暴露：
 * - [getBondedDevices] 系统已配对设备列表
 * - [connect] / [disconnect] 建立/断开 SPP 通道
 * - [send] 发送一条协议指令（move/home/stop/get_version…）
 * - [onFrameReceived] 接收回调（设备回的 ack/status/version…）
 */
class BaseService {

    companion object {
        private const val TAG = "BaseService"
        private val SPP_UUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34AF")
    }

    enum class ConnectionState { DISCONNECTED, CONNECTING, CONNECTED, ERROR }

    data class BaseStatus(
        val dof: Int = 2, val mode: String = "idle",
        val yaw: Double = 0.0, val pitch: Double = 0.0, val roll: Double = 0.0,
    )

    @Volatile var state: ConnectionState = ConnectionState.DISCONNECTED
        private set
    @Volatile var status: BaseStatus = BaseStatus()
        private set
    @Volatile var device: BluetoothDevice? = null
        private set

    val isConnected: Boolean get() = state == ConnectionState.CONNECTED

    /** 收到完整帧（JSON 文本）时的回调。 */
    var onFrameReceived: ((JSONObject) -> Unit)? = null

    private val protocol = BaseProtocol()
    private var socket: BluetoothSocket? = null
    private var readThread: Thread? = null
    @Volatile private var shouldRead = false

    /** 获取本机已配对的经典蓝牙设备列表。 */
    @SuppressLint("MissingPermission")
    fun getBondedDevices(context: Context): List<BluetoothDevice> {
        return try {
            val bm = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
            val adapter = bm.adapter ?: return emptyList()
            adapter.bondedDevices.toList()
        } catch (e: Exception) {
            Log.e(TAG, "getBondedDevices: ${e.message}")
            emptyList()
        }
    }

    /** 连接到指定设备（建立 SPP/RFCOMM 通道）。 */
    @SuppressLint("MissingPermission")
    fun connect(device: BluetoothDevice): Boolean {
        if (state == ConnectionState.CONNECTING) return false
        this.device = device
        state = ConnectionState.CONNECTING
        return try {
            val s = device.createRfcommSocketToServiceRecord(SPP_UUID)
            s.connect()
            socket = s
            protocol.reset()
            state = ConnectionState.CONNECTED
            startReading()
            Log.i(TAG, "已连接 ${device.name ?: device.address}")
            true
        } catch (e: IOException) {
            Log.e(TAG, "连接失败: ${e.message}")
            state = ConnectionState.ERROR
            false
        }
    }

    /** 启动后台读取线程：字节流 → 协议切帧 → 回调。 */
    private fun startReading() {
        shouldRead = true
        readThread = Thread({
            val s = socket ?: return@Thread
            val input = try { s.inputStream } catch (_: Exception) { return@Thread }
            val buf = ByteArray(1024)
            while (shouldRead && s.isConnected) {
                try {
                    val n = input.read(buf)
                    if (n > 0) {
                        val frames = protocol.feed(buf.copyOf(n))
                        for (text in frames) {
                            try {
                                val obj = JSONObject(text)
                                handleFrame(obj)
                                onFrameReceived?.invoke(obj)
                            } catch (_: Exception) {
                                // 坏帧跳过。
                            }
                        }
                    }
                } catch (_: IOException) {
                    break // 连接断开。
                }
            }
            if (shouldRead) handleDisconnect("读取异常断开")
        }, "xbot-base-read").apply { isDaemon = true; start() }
    }

    /** 处理一帧：解析 status/version 更新本地状态。 */
    private fun handleFrame(obj: JSONObject) {
        val cmd = obj.optString("cmd")
        when (cmd) {
            "status" -> status = BaseStatus(
                dof = obj.optInt("dof", status.dof),
                mode = obj.optString("mode", status.mode),
                yaw = obj.optDouble("yaw", status.yaw),
                pitch = obj.optDouble("pitch", status.pitch),
                roll = obj.optDouble("roll", status.roll),
            )
        }
    }

    /** 发送一条协议指令。 */
    @SuppressLint("MissingPermission")
    fun send(cmd: String, params: Map<String, Any?> = emptyMap()): Boolean {
        val s = socket ?: return false
        if (!isConnected) return false
        val bytes = protocol.encode(cmd, params)
        return try {
            s.outputStream.write(bytes)
            s.outputStream.flush()
            true
        } catch (e: IOException) {
            Log.e(TAG, "发送失败: ${e.message}")
            handleDisconnect("发送失败")
            false
        }
    }

    private fun handleDisconnect(reason: String) {
        shouldRead = false
        try { socket?.close() } catch (_: Exception) {}
        socket = null
        state = ConnectionState.DISCONNECTED
        Log.i(TAG, "断开: $reason")
    }

    fun disconnect() {
        handleDisconnect("主动断开")
    }

    fun dispose() {
        disconnect()
    }
}
